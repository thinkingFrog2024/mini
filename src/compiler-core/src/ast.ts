export const enum NodeTypes{
    INTERPOLATION,
    SYMPLE_EXPRESSION,
    ELEMENT,
    TEXT,
    ROOT
}