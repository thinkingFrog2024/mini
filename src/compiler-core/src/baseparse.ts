import { NodeTypes } from "./ast"

const enum TagType{
    Start,
    End
}

export function baseParse(content:string){
    const context = createParserContext(content)

    return createRoot(parseChildren(context,''))
}

function parseChildren(context,parentTag){
    const nodes:Array<any>= []
    let node
    while(!isEnd(context,parentTag)){
        const s = context.source
        if(s.startsWith("{{")){
            node = parseInterpolation(context)
        }else if(s[0]=='<'){
            // 判断是不是元素类型
            if(/[a-z]/i.test(s[1])){
                
                node = pareseElement(context)
            }
        }
        // 如果既不是element类型 也不是插值类型 就当作Text类型处理
        if(!node){
            node = parseText(context)
        }
        nodes.push(node)
    }
    return nodes
}

function isEnd(context,parentTag){
    // 当source没有值   或者需要结束标签</>的时候 停止循环
    const s = context.source
    
    if(parentTag&&s.startsWith("</"+parentTag+'>')){
        return true
    }
    return !s
}

function parseText(context){

    // 遇到插值语法的时候应该停止截取
    let endIndex = context.source.length
    const endToken = "{{"
    const index = context.source.indexOf(endToken)
    if(index!=-1){
        endIndex = index
    }


    const content = parseTextData(context,endIndex)
    adviceBy(context,content.length)
    return{
        type:NodeTypes.TEXT,
        content:content
    }

}


function parseTextData(context,length){
    return context.source.slice(0,length)
}

function pareseElement(context){
    // 解析tag
    // 删除处理完成的代码
    // 这个正则表达式括号里面是捕获内容
    // 返回的数组里面的第一个元素是整个匹配到的内容 在这里是<div
    // 之后的内容是捕获组捕获到的内容 也就是div

    // 处理<div>
    const element:any = parseTag(context,TagType.Start)
    // 处理element里面的children 经过parseElement的处理element开头的符号已经被去掉了
    element.children = parseChildren(context,element.tag)

    // 处理</div>
    parseTag(context,TagType.End)

    return element
}

function parseTag(context,type){
    const match:any = /^<\/?([a-z]*)/i.exec(context.source)
    const tag = match[1]
    adviceBy(context,match[0].length)
    adviceBy(context,1)
    if(type === TagType.Start){
        return {
            type:NodeTypes.ELEMENT,
            tag:tag
        }
    }
}

function parseInterpolation(context){
    console.log('_________________________________');
    
// {{message}}=>message
    const openDelemeter = "{{"
    const closeDelimeter = "}}"


    const closeIndex = context.source.indexOf(closeDelimeter,openDelemeter.length)
    console.log(closeIndex);
    
    adviceBy(context,openDelemeter.length)
    
    const rawContentLength = closeIndex - openDelemeter.length
    console.log(rawContentLength);
    
    const rawcontent = parseTextData(context,rawContentLength)
    console.log(rawContentLength);
    
    // 删除已经处理的部分
    adviceBy(context,rawContentLength+closeDelimeter.length)
    const content = rawcontent.trim()
    return {
        type:NodeTypes.INTERPOLATION,
        content:{
            type:NodeTypes.SYMPLE_EXPRESSION,
            content:content
        }
    }
}

function createRoot(child){
    return {
        children:child
    }
}

function adviceBy(context:any,length:number){
    context.source = context.source.slice(length)
}

function createParserContext(content){
    return{
        source:content
    }
}