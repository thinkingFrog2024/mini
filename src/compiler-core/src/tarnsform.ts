import { NodeTypes } from "./ast";

export function transform(root,options){
    // 遍历 修改textContent
    console.log(options);
    
    const context = createTransformContext(root,options)

    tranverseNode(root,context)
    createRootcodegen(root)
    root.helpers = [...context.helpers.keys()]
}

function createRootcodegen(root){
    root.codegenNode = root.children[0]
}

function tranverseNode(node,context){
    console.log(node);
    const nodeTransform = context.options.nodeTransforms
    if(nodeTransform){
        for(let i = 0;i<nodeTransform.length;i++){
            nodeTransform[i](node)
        }
    }

    const {children,type} = node

    switch(type){
        case NodeTypes.INTERPOLATION:
            context.helper("toDisplayString")
            break
        case NodeTypes.ELEMENT:
        case NodeTypes.ROOT:
            tranverseChildren(children,context)
            
        default:
            break
    }

}  

function tranverseChildren(children,context){
    for(let i = 0;i<children.length;i++){
        const node = children[i]
        tranverseNode(node,context)
    }
}


function createTransformContext(root,options){
    const context = {
        root,
        options,
        helper(key){
            context.helpers.set(key,1)
        },
        helpers:new Map
    }
    return context
}

