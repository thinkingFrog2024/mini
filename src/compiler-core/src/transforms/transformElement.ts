import { NodeTypes } from "../ast";
import { CREATEELEMENTVNODE } from "./runtimeHelp";

// 在ast的helpers数组里面添加上CREATEELEMENTVNODE 实现导入逻辑
export function trandformElement(node,context){
    if(node.type === NodeTypes.ELEMENT){
        context.helper(CREATEELEMENTVNODE)
        node.tag = `'${node.tag}'`
    }
}