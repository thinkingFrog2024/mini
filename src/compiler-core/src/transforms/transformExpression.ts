import { NodeTypes } from "../ast";

export function transformExpression(node){
    // 插值类型=>content=>字符串类型
    if(node.type === NodeTypes.INTERPOLATION){
        const raw = node.content.content
        node.content.content = '_ctx.'+raw
    }
}