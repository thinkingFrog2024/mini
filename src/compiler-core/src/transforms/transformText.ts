import { NodeTypes } from "../ast";

export function transformText(node){
    if(node.type == NodeTypes.ELEMENT){
        // 遍历元素类型的子节点 并且判断子节点的类型
        const {children} = node
        let Container
        for(let i = 0;i<children.length;i++){
            const child = children[i]
            
            if(isText(children[i])){
                // 遍历这个节点以后的节点 判断是不是text 如果相邻的节点也是text 那么就可以初始化一个容器
                // 这个容器实际上也是一个节点 把后面的所有text类型的节点都添加到这个节点的children里面
                // 并且把添加的节点从原来的children里面删掉
                for(let j = i+1;j<children.length;j++){
                    const next = children[j]
                    if(isText(next)){
                        // 初始化容器 并且把children[i]赋值成这个容器
                        if(!Container){
                            Container = children[i] = {
                                type:NodeTypes.COMPOUND,
                                children:[child]
                            }
                        }
                        // 添加加号 方便拼接
                        Container.children.push('+')
                        // 向容器添加节点
                        Container.children.push(next)
                        // 删除children里面的节点
                        children.splice(j,1)
                        
                        // 由于数组里面进行删除 元素减少了 j需要减一
                        j--
                    }else{
                        // 如果下一个元素不是text bane就应该重置容器 离开当前的循环
                        Container.children.push(',')
                        Container = null
                        break
                    }
                }
            }
        }
    }
}

function isText(node){
    return node.type ===NodeTypes.TEXT||node.type === NodeTypes.INTERPOLATION
}