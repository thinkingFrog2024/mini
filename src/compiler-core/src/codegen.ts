import { NodeTypes } from "./ast";
import { CREATEELEMENTVNODE, helpMapName, TODISPLAYSTRING } from "./transforms/runtimeHelp";


// 插值生成render函数相比于字符串生成render函数多了一个导入逻辑 返回数据的时候是通过函数调用:toDisplayString(因为从代理对象里面取出的值未必是一个字符串)

export function generate(ast){
    console.log(ast);
    
    const context = createCodegenContext() //前导函数
    const {push} = context
    getFunctionPreamble(ast,context)
    const functionName = 'render' //函数名
    const args = ['_ctx','_cache'] //参数数组
    const signature = args.join(',') //数组转成string
    push(`function ${functionName}(${signature}){`) //拼接 
    push('return')
    getNode(ast.codegenNode,context) //处理不同类型的节点
    push('}')

    return {
        code:context.code
    } //返回code
}

function createCodegenContext(){
    const context = {
        code:"",
        push(str){
            context.code+=str
        },
        helper(key){
            return `_${helpMapName[key]}`
        }
    }
    return context
}

function getNode(node,context){
    // 处理不同的数据类型
    switch(node.type){
        case NodeTypes.TEXT:
            genText(node,context)
            break
        case NodeTypes.ELEMENT:
            genElement(node,context)
            break
        case NodeTypes.INTERPOLATION:
            console.log(node)
            
            genInterpolation(node,context)
            break
        case NodeTypes.SYMPLE_EXPRESSION:
            genExpression(node,context)
            break
        case NodeTypes.COMPOUND:
            genCompound(node,context)
            break
    }
    
}

function genCompound(node,context){
    const {children} = node
    const {push} = context
    for(let i = 0;i<children.length;i++){
        const child = children[i]
        if(typeof child === 'string'){
            push(child)
        }else{
            getNode(child,context)
        }
    }
}

function genElement(node,context){
    const {push,helper} = context
    const {tag,children} = node
    push(`${helper(CREATEELEMENTVNODE)}(${tag}),null`)
    // push(`${helper(CREATEELEMENTVNODE)}(${tag}),null,"hi," + _toDisplayString(_ctx.message)}`)
    // 处理子节点
    for(let i = 0;i<children.length;i++){
        const child = children[i]
        getNode(child,context)
    }
}

function genExpression(node,context){
    const {push} = context
    // 这个应该放在transform里面进行处理
    // push('_ctx.')
    push(node.content)
}

function genText(node,context){
    const {push} = context
    push(`'${node.content}'`)
}

function genInterpolation(node,context){
    const {push,helper} = context
    push(`${helper(TODISPLAYSTRING)}(`)
    console.log(node.content);
    
    getNode(node.content,context)
    push(')')
}

function getFunctionPreamble(ast,context){
    const {push} = context
    const vueBinging = 'Vue'
    // 把helpers里面的Symbol映射处理成字符串
    // 处理导入逻辑
    const aliasHelpers = (s)=>`${helpMapName[s]}:_${helpMapName[s]}`
    if(ast.helpers.length>0){
        push(`const {${ast.helpers.map(aliasHelpers).join(',')} = ${vueBinging}}`)
    }
    push('\n')
    push('return  ')
}