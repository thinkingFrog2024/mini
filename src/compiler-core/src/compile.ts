import { baseParse } from "../src/baseparse"
import { generate } from "../src/codegen"
import { transform } from "../src/tarnsform"
import { trandformElement } from "../src/transforms/transformElement"
import { transformExpression } from "../src/transforms/transformExpression"
import { transformText } from "../src/transforms/transformText"


export function basecompile(template){
    const ast:any = baseParse(template)
    transform(ast,{
        nodeTransforms:[transformExpression,trandformElement,transformText]
    })
    const {code} = generate(ast)
    return code
}
// const Vue:any = {}
// const {createElementVNode:_createElementVNode,toDisplayString:_toDisplayString} = Vue
// function render(_ctx,_cache){
// return _createElementVNode('div'),null,'hi, '+_toDisplayString(_ctx.message)+_toDisplayString(_ctx.xixi)+'nini',_createElementVNode('span'),null}