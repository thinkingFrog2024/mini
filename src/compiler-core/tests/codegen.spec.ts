import { baseParse } from "../src/baseparse"
import { generate } from "../src/codegen"
import { transform } from "../src/tarnsform"
import { trandformElement } from "../src/transforms/transformElement"
import { transformExpression } from "../src/transforms/transformExpression"
import { transformText } from "../src/transforms/transformText"

describe('生成render函数',()=>{
    it('字符串生成render函数',()=>{
        const ast = baseParse('string')
        transform(ast,{})
        const {code} = generate(ast)
        expect(code).toMatchSnapshot()
    })
    it('插值生成render函数',()=>{
        const ast = baseParse('{{message}}')
        // 这个插件用于在变量名也就是message前面添加_ctx.
        transform(ast,{
            nodeTransforms:[transformExpression]
        })
        const {code} = generate(ast)
        expect(code).toMatchSnapshot()
    })
    // 给element类型添加插件
    it('element生成render函数',()=>{
        const ast = baseParse('<div></div>')
        transform(ast,{
            nodeTransforms:[trandformElement]
        })
        const {code} = generate(ast)
        expect(code).toMatchSnapshot()
    })
    it('联合类型生成render函数',()=>{
        const ast:any = baseParse('<div>hi, {{message}}{{xixi}}nini<span></span></div>')
        transform(ast,{
            // 这里虽然调用了trandformElement插件 但是得到的结果里里面的插值表达式前面并没有_ctx
            // 这是因为:插件是对根节点的每个子节点调用的 而ast的结构是:根节点=>元素节点=>文本节点 插值节点
            // 最初对元素节点调用transformExpression 改变了元素节点的childeren ast结构变成了元素节点=>复合节点
            // 但是原来的代码只会遍历root element的子节点
            // 所以我觉得这里在switch里面对复合类型也调用就可以解决这个问题
            // 但是视频里面是修改了调用顺序 在ast的结构被修改之前调用插件

            // ele节点使用transformExpression =》 类型不匹配 下一个
            // 使用trandformElement =》 类型匹配 修改了ast结构
            // 解决方案=》transformExpression先对每个节点进行使用 另外两个插件在这个插件运行完了在使用
            // 初始化一个数组存放后面调用的函数 修改trandformElement,transformText]使其返回一个函数执行原来的逻辑
            // 把他们返回的函数存放在数组里面 在tranverseNode函数的末尾依次调用 这样就实现了对每个节点先调用transformExpression
            // 把进入执行=》退出执行
            nodeTransforms:[transformExpression,trandformElement,transformText]
        })
        console.log(ast.codegenNode.children);
        
        const {code} = generate(ast)
        expect(code).toMatchSnapshot()
    })
}) 

// 联合类型处理 生成的render函数里面会有一个div 第二个参数则是props 第三个参数则是子节点
