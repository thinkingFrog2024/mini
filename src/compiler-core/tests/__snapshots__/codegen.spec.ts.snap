// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`生成render函数 element生成render函数 1`] = `
"const {createElementVNode:_createElementVNode} = Vue
return  function render(_ctx,_cache){
return _createElementVNode('div',null,[])}"
`;

exports[`生成render函数 字符串生成render函数 1`] = `
"
return  function render(_ctx,_cache){
return 'string'}"
`;

exports[`生成render函数 插值生成render函数 1`] = `
"const {toDisplayString:_toDisplayString} = Vue
return  function render(_ctx,_cache){
return _toDisplayString(_ctx.message)}"
`;

exports[`生成render函数 联合类型生成render函数 1`] = `
"const {createElementVNode:_createElementVNode,toDisplayString:_toDisplayString} = Vue
return  function render(_ctx,_cache){
return _createElementVNode('div',null,['hi, '+_toDisplayString(_ctx.message)+_toDisplayString(_ctx.xixi)+'nini',_createElementVNode('span',null,[])])}"
`;
